hazelcast/src/main/java/com/hazelcast/client/impl/connection/nio/ClientConnectionManagerImpl.java

```diff
@@ -479,278 +682,205 @@ public void onError(Channel channel, Throwable cause) {
-                //this means that server is too old and failover not supported
-                //IllegalStateException will cause client to give up trying on this cluster
-                onFailure(new ClientNotAllowedInClusterException("Cluster does not support failover. "
-                        + "This feature is available in Hazelcast Enterprise with version 3.12 and after"));
-                return false;
-            }
-
-            try {
-                boolean isAllowed = ClientIsFailoverSupportedCodec.decodeResponse(isFailoverFuture.get()).response;
-                if (!isAllowed) {
-                    //in this path server is new but not enterprise.
-                    //IllegalStateException will cause client to give up trying on this cluster
-                    onFailure(new ClientNotAllowedInClusterException("Cluster does not support failover. "
-                            + "This feature is available in Hazelcast Enterprise"));
-                    return false;
-                }
-                return true;
-            } catch (Exception e) {
-                //server version is 3.12 or newer, but an exception occured while getting the isFailover supported info
-                //Exception is delegated without wrapping `IllegalStateException` to so that we will continue
-                // trying same connection
-                onFailure(e);
-                return false;
+    private boolean checkFailoverSupport(Connection connection) {
+        if (!failoverConfigProvided) {
+            return true;
+        }
+        ClientMessage isFailoverSupportedMessage = ClientIsFailoverSupportedCodec.encodeRequest();
+        ClientInvocationFuture future = new ClientInvocation(client, isFailoverSupportedMessage, null, connection).invoke();
+        try {
+            boolean isAllowed = ClientIsFailoverSupportedCodec.decodeResponse(future.get()).response;
+            if (!isAllowed) {
+                logger.warning("Cluster does not support failover. "
+                        + "This feature is available in Hazelcast Enterprise");
             }
+            return isAllowed;
+        } catch (InterruptedException | ExecutionException e) {
+            logger.warning("Cluster did not answer to failover support query. ", e);
+            return false;
         }
+    }
 
-        private void onAuthenticated() {
-            Address memberAddress = connection.getEndPoint();
-            ClientConnection oldConnection = activeConnections.put(inetSocketAddressCache.get(memberAddress), connection);
-            if (oldConnection == null) {
-                if (logger.isFinestEnabled()) {
-                    logger.finest("Authentication succeeded for " + connection
-                            + " and there was no old connection to this end-point");
+    private class OpenConnectionToCluster implements Runnable {
```

Comment:
Class name is misleading as if it is the first cluster connection. Can we change to something like `tryConnectingToAllMembers` or something like this?